plaid/src/data/websocket/mod.rs

```diff
@@ -0,0 +1,408 @@
+fn parse_uris<'de, D>(deserializer: D) -> Result<Vec<Uri>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let uris_raw = <Vec<String>>::deserialize(deserializer)?;
+
+    let uris = uris_raw
+        .iter()
+        .filter_map(|uri| match Uri::from_str(uri) {
+            Ok(valid_uri) => Some(valid_uri),
+            Err(e) => {
+                error!("Invalid URI provided: {}. Error: {}", uri, e);
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if uris.is_empty() {
+        Err(serde::de::Error::custom(&format!("No valid URIs provided")))
+    } else {
+        Ok(uris)
+    }
+}
+
+/// Custom parser to convert user provided `HashMap<String, String>` of headers to include in the request
+/// to a `http::HeaderMap`. Returns an error if the conversion to `HeaderMap` fails.
+fn parse_headers<'de, D>(deserializer: D) -> Result<Option<HeaderMap>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let headers: Option<HashMap<String, String>> = Option::deserialize(deserializer)?;
+
+    if let Some(ref headers) = headers {
+        match headers.try_into() {
+            Ok(map) => Ok(Some(map)),
+            Err(e) => Err(serde::de::Error::custom(format!(
+                "Invalid headers provided: {e}"
+            ))),
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+/// A generator that manages multiple WebSocket clients for data generation.
+///
+/// The `WebsocketGenerator` struct holds a collection of `WebSocketClient` instances and provides
+/// methods to initialize and start these clients.
+pub struct WebsocketGenerator {
+    clients: Vec<WebSocketClient>,
+}
+
+/// Creates a new `WebsocketGenerator` instance with the specified configuration and message sender.
+///
+/// This function initializes the WebSocket clients based on the provided configuration and returns
+/// a `WebsocketGenerator` instance containing these clients.
+///
+/// # Parameters
+/// - `config`: The configuration for the WebSocket data generator, containing a list of WebSocket
+///   configurations.
+/// - `sender`: A channel sender for sending messages to the executor.
+///
+/// # Returns
+/// A new `WebsocketGenerator` instance.
+impl WebsocketGenerator {
+    pub fn new(config: WebsocketDataGenerator, sender: Sender<Message>) -> Self {
+        let clients = config
+            .websockets
+            .into_iter()
+            .map(|(name, config)| WebSocketClient::new(config, sender.clone(), name))
+            .collect();
+
+        Self { clients }
+    }
+
+    /// Starts all WebSocket clients managed by this generator.
+    ///
+    /// This function initializes the WebSocket clients, logs the number of clients being initialized,
+    /// and then spawns an asynchronous task for each client. Each task runs in a loop, attempting to
+    /// start the client and reopening the connection with a new URI if an error occurs.
+    pub async fn start(self) {
```

Comment:
Is it intended you want this to consume itself? It's not a problem, just making sure that was an intentional decision.

Reply:
Yeah it's intentional. The associated methods on the socket require a mutable reference so I have the start method consume itself